### ISSUE TITLE
Upload files with changed `params` but unchanged content

### ISSUE THREAD

**Issue by ChrisWren** [MathieuLoutre/grunt-aws-s3#31]
I love using the differential feature to save time by only uploading modified assets. However, if I change the caching header or content encoding of a resource, and the content stays the same, the resource will not be re-uploaded with the new headers.

Maybe we could add another config property to re-upload files if any headers change? Since `differential` is already used as a part of the aws node sdk we probably wouldn't want to change how that works. Solving this issue would be done by implementing a helper property like `uploadChangedHeaders` to solve the problem described above.

Any thoughts?

**Comment by dedsm** [comment]
I was thinking on this, how about having a data file uploaded with all the metadata of everything uploaded and use that file to do these comparisons?

**Comment by MathieuLoutre** [comment]
@dedsm That's an idea, however I've tried to stay away from manifest-type files because they're a pain to maintain. 

I'm still stuck on this problem since `listObjects` doesn't include the header info. Because changing the headers iis not likely to be a repeat operation (unless I'm missing something), one way of doing it would be to disable `differential` just to change the headers.

**Comment by dedsm** [comment]
yeah @MathieuLoutre but that's a pain if you are uploading a ton of files, or huge files I think, maybe separating the upload and the header setting when differential is enabled? but I don't know if that's possible with the s3 sdk.

**Comment by MathieuLoutre** [comment]
Guess you're right. Separating setting the headers and uploading the file is an interesting approach but as far as I can read, the Node SDK doesn't offer the option to just set the headers. Arguably, I haven't fully read the docs, so maybe there's a way to use `putObject` without uploading a file and modifying an existing one.

**Comment by dedsm** [comment]
I just tested using putObject without specifying the Body parameter as it was the fastest thing to test, of course it didn't work because it would have been too easy.

Nevertheless, we can use copyObject for what we want, in http://docs.aws.amazon.com/AmazonS3/latest/dev/CopyingObjectsExamples.html says explicitly that you could use the [copyObject](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#copyObject-property) call to change the object metadata by setting the source and target pointing to the same object, I'm writing some code to test this now, but we still have one thing to cope with, I guess issuing a copy operation for _every_ file to change the metadata is out of the question, and we don't know if we have to change the metadata unless we send a [headObject](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#headObject-property) call for each file, that also seems an overkill and a waste of resources for something that's not a common case.

So my proposal is maybe read from a grunt option a special parameter and if it exists issue the copy to every file? something like `grunt aws_s3 --newHeaders`

**Comment by MathieuLoutre** [comment]
Oh, nice find! Thanks for looking into this.

I'm happy with an option. Would probably put it in the grunt config of the files as another option in the style of `differential`. When the option is enabled it triggers a copy call (same destination, new headers) instead of a normal upload.

Could do the trick. What do you think?

**Comment by dedsm** [comment]
I'm more inclined to a command line option, since you don't have to modify the Gruntfile and that'll probably be something you don't use a lot, and when you do it'll be in one run, I don't know what you think

**Comment by MathieuLoutre** [comment]
I wasn't sure it would be possible but I just checked the docs and it is (http://gruntjs.com/api/grunt.option) so, probably a better choice indeed.

**Comment by dedsm** [comment]
ok, I can probably work on this sometime this week, some easy changes, not really sure on testing though

**Comment by MathieuLoutre** [comment]
The current testing relies on `mock-aws-s3` but the module doesn't really set headers. Not sure what's the best way right now.
